std/string.d

```diff
@@ -1897,46 +1897,78 @@ S capitalize(S)(S s) @trusted pure
     $(D "\r\n"), $(XREF uni, lineSep), and $(XREF uni, paraSep) as delimiters.
     If $(D keepTerm) is set to $(D KeepTerminator.yes), then the delimiter
     is included in the strings returned.
-  +/
+
+  Params:
+    s a string of $(D chars), $(D wchars), or $(D dchars)
+    keepTerm whether delimiter is included or not in the results
+  Returns:
+    array of strings, each element is a line that is a slice of $(D s)
+ +/
 alias KeepTerminator = Flag!"keepTerminator";
 /// ditto
 S[] splitLines(S)(S s, in KeepTerminator keepTerm = KeepTerminator.no) @safe pure
     if (isSomeString!S)
 {
-    import std.utf : decode;
     import std.uni : lineSep, paraSep;
     import std.array : appender;
 
     size_t iStart = 0;
-    size_t nextI = 0;
     auto retval = appender!(S[])();
 
-    for (size_t i; i < s.length; i = nextI)
+    for (size_t i; i < s.length; ++i)
     {
-        immutable c = decode(s, nextI);
-
-        if (c == '\r' || c == '\n' || c == lineSep || c == paraSep)
+        switch (s[i])
         {
-            immutable isWinEOL = c == '\r' && i + 1 < s.length && s[i + 1] == '\n';
-            auto iEnd = i;
+            case '\n':
+                retval.put(s[iStart .. i + (keepTerm == KeepTerminator.yes)]);
+                iStart = i + 1;
+                break;
+
+            case '\r':
+                if (i + 1 < s.length && s[i + 1] == '\n')
+                {
+                    retval.put(s[iStart .. i + (keepTerm == KeepTerminator.yes) * 2]);
+                    iStart = i + 2;
+                    ++i;
+                }
+                else
+                {
+                    goto case '\n';
+                }
+                break;
 
-            if (keepTerm == KeepTerminator.yes)
+            static if (s[i].sizeof == 1)
             {
-                iEnd = isWinEOL? nextI + 1 : nextI;
+                /* Manually decode:
+                 *  lineSep is E2 80 A8
+                 *  paraSep is E2 80 A9
+                 */
+                case 0xE2:
+                    if (i + 2 < s.length &&
+                        s[i + 1] == 0x80 &&
+                        (s[i + 2] == 0xA8 || s[i + 2] == 0xA9)
```

Comment:
I'm afraid that the way among() is implemented it won't be inlined or optimized well.